### PR TITLE
Rewrite `core-no-oom-handling`, `issue-24445` and `issue-38237` `run-make` tests to new `rmake.rs` format

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

The test which is now called `non-pie-thread-local` has an unexplained "only-linux" flag. Could it be worth trying to remove it and changing the CI to test non-Linux platforms on it?

### CODE REVIEW

File: tests/run-make/core-no-oom-handling/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// This test checks that the core library can still compile correctly
+// when the no_global_oom_handling feature is turned on.
```
Comment (jieyouxu) [inline tests/run-make/core-no-oom-handling/rmake.rs:2]: Nit: it only tests that *compilation succeeds*, it does not check for *correctness*.

File: tests/run-make/deref-impl-rustdoc-ice/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// A very specific set of circumstances (mainly, implementing Deref, and
+// having a procedural macro and a Debug derivation in external crates) caused
+// an internal compiler error (ICE) when trying to use rustdoc. This test
+// reproduces the exact circumstances which caused the bug and checks
+// that it does not happen again.
```
Comment (jieyouxu) [inline tests/run-make/deref-impl-rustdoc-ice/rmake.rs:5]: Not-a-review-comment: this is *very* specific... it must've been a pain to figure out lol

**Review by jieyouxu — APPROVED** [review]

**Comment by jieyouxu** [comment]
One minor nit for a test comment, then r=me after CI is green
@bors delegate+

File: tests/run-make/core-no-oom-handling/rmake.rs (hunk 1 above)
Comment (Oneirical) [reply tests/run-make/core-no-oom-handling/rmake.rs]: I wrote that exact "correctly" terminology in a couple of similar tests to this one, so I have changed them as well. Thanks!

File: tests/run-make/deref-impl-rustdoc-ice/rmake.rs (hunk 2 above)
Comment (Oneirical) [reply tests/run-make/deref-impl-rustdoc-ice/rmake.rs:5]: I know, right? I find it especially hilarious that the fix for all this, in the linked PR, is literally just adding a single line of code.

File: tests/run-make/non-pie-thread-local/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+fn main() {
+    rustc().input("foo.rs").run();
+    cc().input("foo.c")
+        .arg("-lfoo")
+        .library_search_path(tmp_dir())
+        .arg("-Wl")
+        .arg("--gc-sections")
+        .arg("-lpthread")
+        .arg("-ldl")
+        .out_exe(tmp_dir().join("foo"))
```
Comment (jieyouxu) [inline tests/run-make/non-pie-thread-local/rmake.rs:21]: IIRC `out_exe` tries to construct the executable path under `tmp_dir()` already, so you just need to pass the exe name, i.e. `"foo"` here.

**Review by jieyouxu — APPROVED** [review]

**Comment by jieyouxu** [comment]
Thanks, r=me after CI is green
@bors delegate+ rollup

**Comment by Oneirical** [comment]
```
cc: error: unrecognized command-line option ‘-Wl’; did you mean ‘-W’?
cc: error: unrecognized command-line option ‘--gc-sections’; did you mean ‘--data-sections’?
```

Well, that's an unexpected error. Here is what I think I understand: there are multiple C compilers, and we want to use GCC, as it is the one that has the `-Wl` and `--gc-sections` flags. However, a different compiler was used in the CI, such as Clang?

I also notice how this test has `only-linux` activated, which may indicate that it wants some specificity.

Is it possible that, in the past where this test was written, the GNU/Linux CI was using GCC by default, and that changed? But then, how could the Makefile version of the test keep passing?

I may need your help on this one.

**Comment by jieyouxu** [comment]
> Well, that's an unexpected error. Here is what I think I understand: there are multiple C compilers, and we want to use GCC, as it is the one that has the `-Wl` and `--gc-sections` flags. However, a different compiler was used in the CI, such as Clang?

Not quite: clang and gcc has similar CLI interfaces on this one (e.g. https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-Wl-args). The catch here is that the cli flag takes the form

```bash
$CC -Wl,<args> // <- notice the comma `,` and `<args>` **immediately** after comma, no whitespace
```

So they're actually taking `<args>` and forwarding it to the linker...

**Comment by Oneirical** [comment]
> You needed to write .arg("-Wl,--gc-sections") instead of .arg("-Wl").arg("--gc-sections").

I think the `run-make` House of Horror just claimed a new guest...

Thank you so much for figuring this out, I was digging a serious rabbit hole in obscure documentation...

**Pull request merged**